### PR TITLE
[FLINK-12167] Reset context classloader in run and getOptimizedPlan methods

### PR DESCRIPTION
**Pull request by zorro786**
## What is the purpose of the change

This pull request fixes the bug where ClusterClient doesn't reset context class loader in two methods.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): don't know
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java [hunk 1]
```diff
@@ -197,26 +197,33 @@ public LeaderConnectionInfo getClusterConnectionInfo() throws LeaderRetrievalExc
...
 
-			return env.getOptimizedPlan(prog);
-		} else {
-			throw new RuntimeException("Couldn't determine program mode.");
+				return env.getOptimizedPlan(prog);
+			} else {
+				throw new RuntimeException("Couldn't determine program mode.");
+			}
+		} finally {
+			if (contextCl != null) {
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: I think we don't need the null check here.

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java [hunk 2]
```diff
@@ -197,26 +197,33 @@ public LeaderConnectionInfo getClusterConnectionInfo() throws LeaderRetrievalExc
 	// ------------------------------------------------------------------------
 
 	public static String getOptimizedPlanAsJson(Optimizer compiler, PackagedProgram prog, int parallelism)
-			throws CompilerException, ProgramInvocationException {
+		throws CompilerException, ProgramInvocationException {
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: Unrelated change. Please revert.

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java [hunk 3]
```diff
@@ -197,26 +197,33 @@ public LeaderConnectionInfo getClusterConnectionInfo() throws LeaderRetrievalExc
...
-		Thread.currentThread().setContextClassLoader(prog.getUserCodeClassLoader());
-		if (prog.isUsingProgramEntryPoint()) {
-			return getOptimizedPlan(compiler, prog.getPlanWithJars(), parallelism);
-		} else if (prog.isUsingInteractiveMode()) {
-			// temporary hack to support the optimizer plan preview
-			OptimizerPlanEnvironment env = new OptimizerPlanEnvironment(compiler);
-			if (parallelism > 0) {
-				env.setParallelism(parallelism);
-			}
+		throws CompilerException, ProgramInvocationException {
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: Same here with the formatting.

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java [hunk 4]
```diff
@@ -246,45 +253,51 @@ public static OptimizedPlan getOptimizedPlan(Optimizer compiler, Plan p, int par
...
-			finally {
-				ContextEnvironment.unsetContext();
+			else {
+				throw new ProgramInvocationException("PackagedProgram does not have a valid invocation mode.");
 			}
 		}
-		else {
-			throw new ProgramInvocationException("PackagedProgram does not have a valid invocation mode.");
+		finally {
+			if (contextCl != null) {
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: I think we don't need the null check here.

**Comment by zorro786** [comment]
Thanks for review and fixing the issues @tillrohrmann 
Tests failing here:

```
==============================================================================
Running 'Kafka 0.10 end-to-end test'
==============================================================================
TEST_DATA_DIR: /home/travis/build/apache/flink/flink-end-to-end-tests/test-scripts/temp-test-directory-00251075286
Flink dist directory: /home/travis/build/apache/flink/flink-dist/target/flink-1.9-SNAPSHOT-bin/flink-1.9-SNAPSHOT
Downloading Kafka from https://archive.apache.org/dist/kafka/0.10.2.0/kafka_2.11-0.10.2.0.tgz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   607  100   607    0     0    753      0 --:--:-- --:--:-- --:--:--   754

gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
[FAIL] Test script contains errors.
Checking for errors...
No errors in log files.
Checking for exceptions...
No exceptions in log files.
Checking for non-empty .out files...
grep: /home/travis/build/apache/flink/flink-dist/target/flink-1.9-SNAPSHOT-bin/flink-1.9-SNAPSHOT/log/*.out: No such file or directory
No non-empty .out files.

[FAIL] 'Kafka 0.10 end-to-end test' failed after 0 minutes and 1 seconds! Test exited with exit code 1

No taskexecutor daemon to stop on host travis-job-4a07ec5f-c3be-4797-999e-9b26b211b8e6.
No standalonesession daemon to stop on host travis-job-4a07ec5f-c3be-4797-999e-9b26b211b8e6.
```
Is there any command to re trigger tests?

**Comment by zorro786** [comment]
retest

**Comment by zorro786** [comment]
test

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java [hunk 5]
```diff
@@ -247,44 +252,48 @@ public static OptimizedPlan getOptimizedPlan(Optimizer compiler, Plan p, int par
...
+					else {
+						// in blocking mode, we execute all Flink jobs contained in the user code and then return here
+						return this.lastJobExecutionResult;
+					}
 				}
-				else {
-					// in blocking mode, we execute all Flink jobs contained in the user code and then return here
-					return this.lastJobExecutionResult;
+				finally {
+				ContextEnvironment.unsetContext();
```
Comment (zorro786) [inline flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: Formatting here. I have a reason to force push now and retrigger tests (:
Hope fully the script to untar is fixed.

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java (hunk 3 above)
Comment (zorro786) [reply flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: Is there any style check plugin per Flink codebase I can use to avoid this? My editor moves throws to new line without double indentation

Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: No, unfortunately not. Ideally we would have an auto formatter configuration which we could simply share in the community. However, the community never managed to agree on a consistent code style.

**Comment by tillrohrmann** [comment]
The failing test case seems to be unrelated. I've rebased onto the latest master which retriggers the build. If Travis gives green light, then I'll merge this PR.
